odoo/addons/base/models/res_users.py

```diff
@@ -496,9 +496,11 @@ def _check_disjoint_groups(self):
 
     def onchange(self, values, field_names, fields_spec):
         # Hacky fix to access fields in `SELF_READABLE_FIELDS` in the onchange logic.
-        # Put field values in the cache.
+        # Put field values in the cache. And bypass access in readable fields.
         if self == self.env.user:
-            [self.sudo()[field_name] for field_name in self.SELF_READABLE_FIELDS]
+            readable_fields = frozenset(self.SELF_READABLE_FIELDS)
+            [self.sudo()[field_name] for field_name in readable_fields]
+            self.env.cr.cache['res.users-force-readable-fields'] = readable_fields
```

Comment:
Shouldn't we test `self._origin == self.env.user` in the override of `_has_field_access` instead?  This "forced" prefetching and cache looks really lousy to me.

Reply:
To me, all of this override looks lousy. But yes, I'll try by checking the origin.